File: packages/javascript/src/network/webSocketClient.ts
import { sleep } from '../helpers.js'
import type { WebSocketRetryOptions } from '../types.js'
import { newWebSocket, WS_STATES, type IsoWS } from './iso-ws.js'

export { WS_STATES } from './iso-ws.js'

export type WebSocketClientOptions = {
  baseUrl: string | URL
  retry: Required<WebSocketRetryOptions>
  timeout: number
}

function matchesCloseCode(code: number, list: (number | [number, number])[]): boolean {
  for (const item of list) {
    if (typeof item === 'number') {
      if (code === item) return true
    } else {
      const [start, end] = item
      if (code >= start && code <= end) return true
    }
  }
  return false
}

function removeWsListeners(ws?: IsoWS | null): void {
  if (!ws) {
    return
  }

  ws.onopen = null
  ws.onerror = null
  ws.onmessage = null
  ws.onclose = null
}

export class WebSocketClient {
  private readonly baseUrl: string | URL
  private readonly retry: Required<WebSocketRetryOptions>
  private readonly timeout: number

  constructor(options: WebSocketClientOptions) {
    this.baseUrl = options.baseUrl
    this.retry = options.retry
    this.timeout = options.timeout

    // Ensure maxAttemptsPerConnection, maxDelay, maxConnections and timeout are non-negative integers
    this.retry.maxAttemptsPerConnection = Math.max(
      0,
      Math.floor(this.retry.maxAttemptsPerConnection)
    )
    this.retry.maxDelay = Math.max(0, Math.floor(this.retry.maxDelay))
    this.retry.maxConnections = Math.max(0, Math.floor(this.retry.maxConnections))
    this.timeout = Math.max(0, Math.floor(this.timeout))
  }

  createSession(url: string): WebSocketSession {
    return new WebSocketSession({
      retry: this.retry,
      timeout: this.timeout,
      url: new URL(url, this.baseUrl),
    })
  }
}

class WebSocketSession implements Omit<IsoWS, 'onopen'> {
  onconnecting: ((event: { connection: number; attempt: number }) => void) | null = null
  onopen: ((event: { connection: number; attempt: number }) => void) | null = null
  onerror: IsoWS['onerror'] = null
  onclose: IsoWS['onclose'] = null
  onmessage: IsoWS['onmessage'] = null

  private _readyState: IsoWS['readyState'] = WS_STATES.CONNECTING
  private _url: IsoWS['url']
  private readonly retry: Required<WebSocketRetryOptions>
  private readonly timeout: number

  private ws: IsoWS | null = null
  private connectionCount = 0
  private connectionAttempt = 0
  private connectionTimeoutId: ReturnType<typeof setTimeout> | undefined

  constructor({
    retry,
    timeout,
    url,
  }: {
    retry: Required<WebSocketRetryOptions>
    timeout: number
    url: string | URL
  }) {
    this._url = url.toString()

    this.retry = retry
    this.timeout = timeout

    this.connect().catch(() => {
      // errors are emitted via event handlers; no throw
    })
  }

  get readyState(): IsoWS['readyState'] {
    return this._readyState
  }

  get url(): IsoWS['url'] {
    return this._url
  }

  send(data: string | ArrayBufferLike | Blob | ArrayBufferView): void {
    if (this.readyState === WS_STATES.OPEN) {
      if (!this.ws) {
        throw new Error('readyState is open but ws is not initialized')
      }
      this.ws.send(data)
    } else {
      throw new Error('WebSocket is not open')
    }
  }

  close(code = 1000, reason = ''): void {
    if (this.readyState === WS_STATES.CLOSING || this.readyState === WS_STATES.CLOSED) {
      return
    }

    this.clearConnectionTimeout()
    this._readyState = WS_STATES.CLOSING

    if (this.ws?.readyState === WS_STATES.OPEN) {
      this.ws.close(code)
    } /* if (this.readyState === WS_STATES.CONNECTING) */ else {
      this.onWsClose(code, reason)
    }
  }

  private onWsClose(code = 1000, reason = ''): void {
    this.clearConnectionTimeout()

    if (this.readyState !== WS_STATES.CLOSED) {
      this._readyState = WS_STATES.CLOSED
      this.onclose?.({ code, reason })
    }

    this.onconnecting = null
    this.onopen = null
    this.onclose = null
    this.onerror = null
    this.onmessage = null

    removeWsListeners(this.ws)
    this.ws = null
  }

  private async connect(isRetry = false): Promise<void> {
    this.clearConnectionTimeout()

    if (!isRetry) {
      this.connectionCount += 1
      this.connectionAttempt = 0
    }
    this.connectionAttempt += 1
    this._readyState = WS_STATES.CONNECTING
    this.onconnecting?.({ connection: this.connectionCount, attempt: this.connectionAttempt })

    if (this.timeout > 0) {
      this.connectionTimeoutId = setTimeout(() => {
        this.close(3008, 'WebSocket connection timeout')
      }, this.timeout)
    }

    const onError = async (ws: IsoWS | null, err?: Error) => {
      this.clearConnectionTimeout()
      removeWsListeners(ws)

      if (this.readyState !== WS_STATES.CONNECTING) {
        return
      }

      const noRetry =
        this.retry.maxAttemptsPerConnection > 0 &&
        this.connectionAttempt >= this.retry.maxAttemptsPerConnection
      if (noRetry) {
        this.onerror?.(new Error('WebSocket connection error', { cause: err }))
        this.close(1006, 'WebSocket connection error')
        return
      }

      const delayMs = Math.min(this.retry.delay(this.connectionAttempt), this.retry.maxDelay)
      await sleep(delayMs)
      if (this.readyState === WS_STATES.CONNECTING) {
        this.connect(true)
      }
    }

    let ws: IsoWS
    try {
      ws = await newWebSocket(this.url)
    } catch (err) {
      onError(null, err instanceof Error ? err : new Error(String(err)))
      return
    }

    if (this.readyState !== WS_STATES.CONNECTING) {
      ws.close(1001)
      return
    }

    ws.onerror = () => onError(ws)
    ws.onopen = () => {
      this.clearConnectionTimeout()
      removeWsListeners(ws)

      if (this.readyState !== WS_STATES.CONNECTING) {
        // User closed the connection during the connection attempt
        ws.close(1001)
        return
      }

      ws.onmessage = (event) => {
        this.onmessage?.(event)
      }
      ws.onclose = (event) => {
        removeWsListeners(ws)

        if (this.ws !== ws) {
          // Should not be possible ?
          return
        }

        this.ws = null

        if (this.readyState === WS_STATES.CLOSING) {
          this.onWsClose(event.code, event.reason || '')
          return
        }

        if (
          (this.retry.maxConnections > 0 && this.connectionCount >= this.retry.maxConnections) ||
          !matchesCloseCode(event.code, this.retry.closeCodes)
        ) {
          this.close(event.code, event.reason || '')
          return
        }

        this.connect()
      }

      this.ws = ws
      this._readyState = WS_STATES.OPEN
      this.onopen?.({ connection: this.connectionCount, attempt: this.connectionAttempt })
    }
  }

  private clearConnectionTimeout(): void {
    if (this.connectionTimeoutId) {
      clearTimeout(this.connectionTimeoutId)
      this.connectionTimeoutId = undefined
    }
  }
}

export type { WebSocketSession }

File: packages/sdk-python/src/gladiaio_sdk/network/websocket_client.py
"""Async WebSocket client/session with retry and timeout semantics matching the JS SDK."""

import threading
import time
from collections.abc import Callable
from typing import Any, final

from websockets import ConnectionClosed
from websockets.sync import client as _ws_client

from gladiaio_sdk.client_options import WebSocketRetryOptions

from .types import WS_STATES


def _matches_close(code: int, rules: list[int | tuple[int, int]] | None) -> bool:
  if not rules:
    return False
  for rule in rules:
    if isinstance(rule, tuple):
      start, end = rule
      if start <= code <= end:
        return True
    else:
      if code == rule:
        return True
  return False


@final
class WebSocketSession:
  onconnecting: Callable[[dict[str, int]], None] | None = None
  onopen: Callable[[dict[str, int]], None] | None = None
  onerror: Callable[[Exception], None] | None = None
  onclose: Callable[[dict[str, Any]], None] | None = None
  onmessage: Callable[[dict[str, Any]], None] | None = None

  _ready_state: WS_STATES = WS_STATES.CONNECTING
  _url: str
  _retry: WebSocketRetryOptions
  _timeout: float
  _ws: _ws_client.ClientConnection | None = None
  _connection_count: int = 0
  _connection_attempt: int = 0
  _thread: threading.Thread | None = None
  _stop: threading.Event
  _send_lock: threading.Lock

  def __init__(
    self, url: str, retry: WebSocketRetryOptions, timeout: float, *, base_url: str
  ) -> None:
    self._url = url
    self._retry = retry
    self._timeout = timeout
    self._stop = threading.Event()
    self._send_lock = threading.Lock()

  @property
  def ready_state(self) -> WS_STATES:
    return self._ready_state

  @property
  def url(self) -> str:
    return self._url

  def send(self, data: str | bytes) -> None:
    if self.ready_state == WS_STATES.OPEN:
      if not self._ws:
        raise RuntimeError("readyState is open but ws is not initialized")
      with self._send_lock:
        self._ws.send(data)
    else:
      raise RuntimeError("WebSocket is not open")

  def close(self, code: int = 1000, reason: str = "") -> None:
    if self.ready_state in (WS_STATES.CLOSING, WS_STATES.CLOSED):
      return

    self._ready_state = WS_STATES.CLOSING
    self._stop.set()

    if self._ws and self._ws.state == 1:
      self._ws.close(code=code)
    else:
      self._on_ws_close(code, reason)

  def start(self) -> None:
    """Start the background receive/reconnect loop in a dedicated thread."""
    if self._thread and self._thread.is_alive():
      return
    t = threading.Thread(target=self._connect, name="ws-recv", daemon=True)
    self._thread = t
    t.start()

  def _connect(self, is_retry: bool = False) -> None:
    if not is_retry:
      self._connection_count += 1
      self._connection_attempt = 0
    self._connection_attempt += 1
    self._ready_state = WS_STATES.CONNECTING
    if self.onconnecting:
      self.onconnecting(
        {
          "connection": self._connection_count,
          "attempt": self._connection_attempt,
        }
      )

    def on_error(err: Exception | None) -> None:
      if self._ready_state != WS_STATES.CONNECTING:
        return
      no_retry = (
        self._retry.max_attempts_per_connection > 0
        and self._connection_attempt >= self._retry.max_attempts_per_connection
      )
      if no_retry:
        if self.onerror:
          self.onerror(Exception("WebSocket connection error" if err is None else str(err)))
        self.close(1006, "WebSocket connection error")
        return
      time.sleep(self._retry.delay(self._connection_attempt))
      if self._ready_state == WS_STATES.CONNECTING:
        self._connect(True)

    try:
      if self._timeout > 0:
        ws = _ws_client.connect(
          self._url,
          open_timeout=self._timeout,
        )
      else:
        ws = _ws_client.connect(self._url, open_timeout=None)
    except Exception as e:
      on_error(e)
      return

    if self._ready_state != WS_STATES.CONNECTING:
      ws.close(code=1001)
      return

    self._ws = ws
    self._ready_state = WS_STATES.OPEN
    if self.onopen:
      self.onopen(
        {
          "connection": self._connection_count,
          "attempt": self._connection_attempt,
        }
      )

    close_code: int = 1006
    close_reason: str = "Abnormal closure"
    try:
      while not self._stop.is_set():
        msg = ws.recv()
        if self.onmessage:
          self.onmessage({"data": msg})
    except ConnectionClosed as e:
      close_code = e.code
      close_reason = e.reason

    if self._ws is not ws:
      return

    self._ws = None

    if self.ready_state == WS_STATES.CLOSING:
      self._on_ws_close(close_code, close_reason)
      return

    if (
      self._retry.max_connections > 0 and self._connection_count >= self._retry.max_connections
    ) or not _matches_close(close_code, self._retry.close_codes):
      self.close(
        close_code,
        close_reason,
      )
      return

    if not self._stop.is_set():
      self._connect(False)

  def _on_ws_close(self, code: int = 1000, reason: str = "") -> None:
    if self._ready_state != WS_STATES.CLOSED:
      self._ready_state = WS_STATES.CLOSED
      if self.onclose:
        self.onclose({"code": code, "reason": reason})

    self.onconnecting = None
    self.onopen = None
    self.onclose = None
    self.onerror = None
    self.onmessage = None
    self._ws = None


@final
class WebSocketClient:
  def __init__(self, base_url: str, retry: WebSocketRetryOptions, timeout: float) -> None:
    self._base_url = base_url
    self._retry = retry
    self._timeout = timeout

  def create_session(self, url: str) -> WebSocketSession:
    # TODO use base_url
    return WebSocketSession(url, self._retry, self._timeout, base_url=self._base_url)
